### ISSUE TITLE
Add AI code for automatically choosing the best professors for defense sessions

### ISSUE THREAD

**Issue by semjz** [semjz/uni_thesis#3]

**Comment by mesk2** [comment]
۱. مدل‌ها (models.py)
ابتدا، باید مدل‌های جنگو را برای Professor و Student تعریف کنید.
# app_name/models.py

from django.db import models
from django.contrib.auth.models import User

class Professor(models.Model):
    user = models.OneToOneField(User, on_delete=models.CASCADE)
    field = models.CharField(max_length=50)
    num_theses = models.IntegerField(default=0)
    num_defended_theses = models.IntegerField(default=0)
    
    def __str__(self):
        return self.user.username

class ProfessorTime(models.Model):
    professor = models.ForeignKey(Professor, on_delete=models.CASCADE, related_name='available_times')
    time_slot = models.CharField(max_length=20)

class Student(models.Model):
    user = models.OneToOneField(User, on_delete=models.CASCADE)
    field = models.CharField(max_length=50)
    thesis_title = models.CharField(max_length=255)
    supervisor = models.ForeignKey(Professor, on_delete=models.SET_NULL, null=True, blank=True)
    
    def __str__(self):
        return self.user.username

. منطق (views.py)
در فایل views.py، توابع مربوط به هوش مصنوعی را فراخوانی کنید.
# app_name/views.py

from django.shortcuts import render, redirect
from django.contrib.auth.decorators import login_required
from django.contrib.auth.forms import UserCreationForm
from django.contrib.auth import login, authenticate
from .models import Professor, Student, ProfessorTime
from .ai_logic import select_committee_logic, train_ai_model # ایمپورت کردن کد هوش مصنوعی

# View برای درخواست دانشجو
@login_required
def student_request_view(request):
    if request.method == 'POST':
        # دریافت اطلاعات از فرم
        thesis_title = request.POST.get('thesis_title')
        field = request.POST.get('field')
        supervisor_id = request.POST.get('supervisor_id')
        
        # به‌روزرسانی اطلاعات دانشجو
        student = request.user.student
        student.thesis_title = thesis_title
        student.field = field
        student.supervisor_id = supervisor_id
        student.save()
        
        # فراخوانی منطق هوش مصنوعی
        student_data = {
            'field': student.field,
            'supervisor_id': student.supervisor.id,
        }
        result = select_committee_logic(student_data)
        
        # نمایش نتیجه به کاربر
        if result['status'] == 'success':
            message = f"کمیته با موفقیت تعیین شد. داور: {result['data']['reviewer_id']}, ناظر: {result['data']['observer_id']}"
            return render(request, 'student_result.html', {'message': message})
        else:
            return render(request, 'student_form.html', {'error': result['message']})

    # ... سایر View ها
    return render(request, 'student_form.html')


# View برای استاد
@login_required
def professor_profile_view(request):
    professor = request.user.professor
    if request.method == 'POST':
        # دریافت زمان‌ها از فرم و ذخیره در دیتابیس
        professor.field = request.POST.get('field')
        professor.save()

        # حذف زمان‌های قبلی و ذخیره زمان‌های جدید
        ProfessorTime.objects.filter(professor=professor).delete()
        selected_times = request.POST.getlist('available_times')
        for time_slot in selected_times:
            ProfessorTime.objects.create(professor=professor, time_slot=time_slot)
        
        # آموزش مجدد مدل پس از تغییر زمان‌های اساتید
        train_ai_model()
        return redirect('professor_profile')
    
    # ... سایر View ها
    return render(request, 'professor_profile.html', {'professor': professor})

**Comment by mesk2** [comment]
اگه مشکلی بود بگو

**Comment by mesk2** [comment]
کد جدید 


import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np

# FIELDS و سایر ثوابت مورد نیاز
FIELDS = ['AI', 'CS', 'Math', 'Physics', 'EE']

class NeuralNet(nn.Module):
    """
    تعریف مدل شبکه عصبی برای پیش‌بینی مناسب بودن کمیته.
    """
    def __init__(self, input_size):
        super(NeuralNet, self).__init__()
        self.fc1 = nn.Linear(input_size, 16)
        self.fc2 = nn.Linear(16, 8)
        self.fc3 = nn.Linear(8, 1)
        self.relu = nn.ReLU()
        self.sigmoid = nn.Sigmoid()
    
    def forward(self, x):
        x = self.relu(self.fc1(x))
        x = self.relu(self.fc2(x))
        x = self.sigmoid(self.fc3(x))
        return x

def train_ai_model(professors_data):
    """
    مدل هوش مصنوعی را با داده‌های ورودی آموزش می‌دهد.

    Args:
        professors_data (dict): دیکشنری حاوی اطلاعات اساتید.
    """
    if len(professors_data) < 2:
        return None, "تعداد اساتید برای آموزش مدل کافی نیست."
    
    samples = []
    labels = []
    all_profs = list(professors_data.values())

    for i in range(len(all_profs)):
        for j in range(i + 1, len(all_profs)):
            p1, p2 = all_profs[i], all_profs[j]
            
            # شرط: رشته تحصیلی یکسان باشد
            if p1['field'] != p2['field']:
                continue
            
            common = len(set(p1['available_times']) & set(p2['available_times']))
            
            # شرط: حداقل یک زمان مشترک وجود داشته باشد
            if common == 0:
                continue
            
            try:
                field_index = FIELDS.index(p1['field'])
                feature = [p1['num_defended_theses'], p2['num_defended_theses'], common, field_index]
                label = int((p1['num_defended_theses'] + p2['num_defended_theses'] < 10) and (common >= 5))
                
                samples.append(feature)
                labels.append(label)
            except ValueError:
                continue
    
    if not samples:
        return None, "داده کافی برای آموزش مدل یافت نشد."
    
    X = torch.tensor(samples, dtype=torch.float32)
    y = torch.tensor(labels, dtype=torch.float32).view(-1, 1)
    
    model = NeuralNet(input_size=4)
    criterion = nn.BCELoss()
    optimizer = optim.Adam(model.parameters(), lr=0.01)
    
    for epoch in range(50):
        model.train()
        optimizer.zero_grad()
        outputs = model(X)
        loss = criterion(outputs, y)
        loss.backward()
        optimizer.step()
    
    model.eval()
    return model, "مدل با موفقیت آموزش دید."


*********************************************************
# selection_logic.py

import torch
import numpy as np
# از ai_model.py ایمپورت کنید
from ai_model import NeuralNet, FIELDS

def select_committee_logic(model, professors_data, student_data):
    """
    کمیته دفاع مناسب را بر اساس داده‌های اساتید و دانشجو انتخاب می‌کند.

    Args:
        model: مدل هوش مصنوعی آموزش‌دیده.
        professors_data (dict): دیکشنری حاوی اطلاعات کامل اساتید.
        student_data (dict): دیکشنری حاوی اطلاعات دانشجو.

    Returns:
        dict: نتیجه انتخاب شامل وضعیت (success/failure) و داده‌های مربوطه.
    """
    if model is None:
        return {"status": "failure", "message": "مدل هوش مصنوعی آماده نیست."}
    
    thesis_field = student_data['field']
    student_supervisor_id = student_data['supervisor_id']
    
    # فیلتر کردن اساتید واجد شرایط
    eligible_profs = [
        p for p in professors_data.values() 
        if p['field'] == thesis_field and 
           p['prof_id'] != student_supervisor_id
    ]
    
    if len(eligible_profs) < 2:
        return {"status": "failure", "message": "تعداد اساتید واجد شرایط کافی نیست."}
    
    features = []
    infos = []

    for i in range(len(eligible_profs)):
        for j in range(i + 1, len(eligible_profs)):
            p1, p2 = eligible_profs[i], eligible_profs[j]
            
            common_times = set(p1['available_times']) & set(p2['available_times'])
            common_count = len(common_times)
            
            if common_count > 0:
                try:
                    field_index = FIELDS.index(p1['field'])
                    f = [p1['num_defended_theses'], p2['num_defended_theses'], common_count, field_index]
                    features.append(f)
                    infos.append({
                        "prof1_id": p1['prof_id'], 
                        "prof2_id": p2['prof_id'], 
                        "common_count": common_count, 
                        "prof1_times": list(p1['available_times']), 
                        "prof2_times": list(p2['available_times'])
                    })
                except ValueError:
                    continue
    
    if not features:
        return {"status": "failure", "message": "هیچ زوج استادی با زمان مشترک پیدا نشد."}
    
    X_new = torch.tensor(features, dtype=torch.float32)
    
    with torch.no_grad():
        preds = model(X_new).numpy().flatten()
    
    best_idx = np.argmax(preds)
    best_info = infos[best_idx]
    best_score = preds[best_idx]
    
    suggested_time = best_info["prof1_times"][0] if best_info["prof1_times"] else "زمان نامشخص"
    
    return {
        "status": "success",
        "data": {
            "reviewer_id": best_info["prof1_id"],
            "observer_id": best_info["prof2_id"],
            "predicted_appropriateness_score": float(best_score),
            "suggested_time_slot": suggested_time,
            "suggested_location": "اتاق سمینار A"
        }
    }***************************************************
نحوه ارتباط با برنامه اصلی

# در برنامه اصلی شما
from ai_model import train_ai_model

# داده‌ها را از دیتابیس بخوانید
all_professors_data = get_professors_data() 

# آموزش مدل
global trained_model
trained_model, message = train_ai_model(all_professors_data)

if trained_model is None:
    print(message) # اگر مدل آموزش ندید
